File: udp/src/main/kotlin/ru/hse/tftp/packet/Packet.kt
package ru.hse.tftp.packet

import ru.hse.tftp.packet.ErrorType.ACCESS_VIOLATION
import ru.hse.tftp.packet.ErrorType.DISK_FULL_OR_ALLOCATION_EXCEEDED
import ru.hse.tftp.packet.ErrorType.FILE_ALREADY_EXISTS
import ru.hse.tftp.packet.ErrorType.FILE_NOT_FOUND
import ru.hse.tftp.packet.ErrorType.ILLEGAL_TFTP_OPERATION
import ru.hse.tftp.packet.ErrorType.NO_SUCH_USER
import ru.hse.tftp.packet.ErrorType.PARSE_ERROR
import ru.hse.tftp.packet.ErrorType.UNKNOWN_TRANSFER_ID
import ru.hse.tftp.packet.MessageType.ACKNOWLEDGEMENT
import ru.hse.tftp.packet.MessageType.DATA
import ru.hse.tftp.packet.MessageType.ERROR
import ru.hse.tftp.packet.MessageType.READ_REQUEST
import ru.hse.tftp.packet.MessageType.WRITE_REQUEST
import ru.hse.tftp.packet.UserRequest.Mode
import ru.hse.tftp.packet.UserRequest.Mode.OCTET
import java.io.ByteArrayOutputStream

enum class MessageType(val value: Int) {
    READ_REQUEST(1),
    WRITE_REQUEST(2),
    DATA(3),
    ACKNOWLEDGEMENT(4),
    ERROR(5),
}

sealed class Packet(val messageType: MessageType) {
    abstract val byteArray: ByteArray
}

private fun ByteArrayOutputStream.writeCode(code: Int) {
    write(code shr 8)
    write(code and 0xff)
}

class ReadPacket(private val filename: String, private val mode: Mode = OCTET): Packet(READ_REQUEST) {
    override val byteArray: ByteArray
        get() {
            val outputStream = ByteArrayOutputStream()
            outputStream.writeCode(messageType.value)
            outputStream.write(filename.toByteArray())
            return outputStream.toByteArray()
        }
}

class WritePacket(private val filename: String, private val mode: Mode = OCTET): Packet(WRITE_REQUEST) {
    override val byteArray: ByteArray
        get() {
            val outputStream = ByteArrayOutputStream()
            outputStream.writeCode(messageType.value)
            outputStream.write(filename.toByteArray())
            return outputStream.toByteArray()
        }
}

class DataPacket(private val data: ByteArray, private val blockNumber: Int): Packet(DATA) {
    override val byteArray: ByteArray
        get() {
            val outputStream = ByteArrayOutputStream()
            outputStream.writeCode(messageType.value)
            outputStream.writeCode(blockNumber)
            outputStream.write(data)
            return outputStream.toByteArray()
        }
}

class AcknolegmentPacket(): Packet(ACKNOWLEDGEMENT) {
    override val byteArray: ByteArray
        get() {
            val outputStream = ByteArrayOutputStream()
            outputStream.writeCode(messageType.value)
            return outputStream.toByteArray()
        }
}

class ErrorPacket(private val type: ErrorType) : Packet(ERROR) {
    val message = when (type) {
        PARSE_ERROR -> "Error parsing request"
        FILE_NOT_FOUND -> "File not found on server"
        ACCESS_VIOLATION -> "Access violation"
        DISK_FULL_OR_ALLOCATION_EXCEEDED -> "Dis is full or allocation exceeded"
        ILLEGAL_TFTP_OPERATION -> "Illegal TFTP operation"
        UNKNOWN_TRANSFER_ID -> "unknown  transfer id"
        FILE_ALREADY_EXISTS -> "File already exists on server"
        NO_SUCH_USER -> "No such user on server"
    }

    override val byteArray: ByteArray
        get() {
            val outputStream = ByteArrayOutputStream()
            outputStream.writeCode(messageType.value)
            outputStream.writeCode(type.code)
            outputStream.write(message.toByteArray())
            return outputStream.toByteArray()
        }
}

File: tftp/packet.go
package main

import (
	"bytes"
	"encoding/binary"
	"errors"
	"strings"
)

//TFTP supports five types of packets, all of which have been mentioned above:
//
//opcode  operation
//  1     Read request (RRQ)
//  2     Write request (WRQ)
//  3     Data (DATA)
//  4     Acknowledgment (ACK)
//  5     Error (ERROR)

type OpCode uint16

const (
	RRQ   OpCode = 1
	WRQ   OpCode = 2
	DATA  OpCode = 3
	ACK   OpCode = 4
	ERROR OpCode = 5
)

func GetOpCode(buf []byte) (*OpCode, error)  {
	var op OpCode
	if err := binary.Read( bytes.NewBuffer(buf), binary.BigEndian, &op); err != nil {
		return nil, err
	}
	switch op {
	case RRQ:
		return &op, nil
	case WRQ:
		return &op, nil
	case DATA:
		return &op, nil
	case ACK:
		return &op, nil
	case ERROR:
		return &op, nil
	default:
		return nil, errors.New("Unknown operation code")
	}
}

type Mode string

const (
	NETASCII Mode = "netascii"
	OCTET    Mode = "octet"
	//MAIL Mode = "mail"
)

func (m Mode) Equals(mode Mode) bool {
	m_ := strings.ToLower(string(m))
	mode_ := strings.ToLower(string(mode))
	return m_ == mode_
}

type Packet interface {
	Parse([]byte) error
}

//        2 bytes    string   1 byte     string   1 byte
//       -----------------------------------------------
//RRQ/  | 01/02 |  Filename  |   0  |    Mode    |   0  |
//WRQ    -----------------------------------------------

type ReadRequestPacket struct {
	Code     OpCode
	Filename string
	EmptyByte1 int8
	Mode     Mode
	EmptyByte2 int8
}

func (p ReadRequestPacket) Parse(buf []byte) error {
	return nil
}

func NewReadPacket(filename string, mode Mode) *ReadRequestPacket {
	return &ReadRequestPacket{
		Code: RRQ,
		Filename: filename,
		EmptyByte1: 0,
		Mode: mode,
		EmptyByte2: 0,
	}
}

type WriteRequestPacket struct {
	Code     OpCode
	Filename string
	EmptyByte1 int8
	Mode     Mode
	EmptyByte2 int8
}

func (p WriteRequestPacket) Parse(buf []byte) error {
	return nil
}

func NewWritePacket(filename string, mode Mode) *WriteRequestPacket {
	return &WriteRequestPacket{
		Code: WRQ,
		Filename: filename,
		EmptyByte1: 0,
		Mode: mode,
		EmptyByte2: 0,
	}
}

//       2 bytes    2 bytes       n bytes
//       ---------------------------------
//DATA  | 03    |   Block #  |    Data    |
//       ---------------------------------

type DataPacket struct {
	Code  OpCode
	Block uint16
	Data  []byte
}

func (p DataPacket) Parse(buf []byte) error {
	return nil
}

func NewDataPacket(block uint16, data []byte) *DataPacket {
	return &DataPacket{
		Code: DATA,
		Block: block,
		Data: data,
	}
}

//       2 bytes    2 bytes
//        -------------------
//ACK   | 04    |   Block #  |
//       --------------------

type AcknowledgementPacket struct {
	Code  OpCode
	Block uint16
}

func (p AcknowledgementPacket) Parse(buf []byte) error {
	return nil
}

func NewAcknowledgementPacket(block uint16) *AcknowledgementPacket {
	return &AcknowledgementPacket{
		Code: ACK,
		Block: block,
	}
}

//Error Codes:
//
//Value     Meaning
//
// 0        Not defined, see error message (if any).
// 1        File not found.
// 2        Access violation.
// 3        Disk full or allocation exceeded.
// 4        Illegal TFTP operation.
// 5        Unknown transfer ID.
// 6        File already exists.
// 7        No such user.

type ErrorCode int

const (
	ErrorNotDefined        ErrorCode = 0
	ErrorFileNotFound      ErrorCode = 1
	ErrorAccessViolation   ErrorCode = 2
	ErrorDiskFill          ErrorCode = 3
	ErrorIllegalOperation  ErrorCode = 4
	ErrorUnknownTid        ErrorCode = 5
	ErrorFileAlreadyExists ErrorCode = 6
	ErrorNoSuchUser        ErrorCode = 7
)

//         2 bytes  2 bytes      string    1 byte
//        ----------------------------------------
// ERROR | 05    |  ErrorCode |   ErrMsg   |   0  |
//        ----------------------------------------

type ErrorPacket struct {
	Code    OpCode
	Error   ErrorCode
	Message string
	EmptyByte int8
}

func (p ErrorPacket) Parse(buf []byte) error {
	return binary.Read(bytes.NewBuffer(buf), binary.BigEndian, p)
}

func NewErrorPacket(code ErrorCode, msg string) *ErrorPacket {
	return &ErrorPacket{
		Code: ERROR,
		Error: code,
		Message: msg,
		EmptyByte: 0,
	}
}
